ledger/ledger-api-bench-tool/src/test/suite/scala/com/daml/ledger/api/benchtool/submission/CommandSubmitterITSpec.scala

```diff
@@ -0,0 +1,307 @@
+    logger = LoggerFactory.getLogger(getClass),
+    expectedTemplateNames = expectedTemplateNames,
+  )
+
+  case class ObservedExerciseEvent(
+      templateName: String,
+      choiceName: String,
+      choiceArgumentsSerializedSize: Int,
+      consuming: Boolean,
+  )
+
+  case class ObservedCreateEvent(templateName: String, createArgumentsSerializedSize: Int)
+
+  case class ObservedEvents(
+      expectedTemplateNames: Set[String],
+      createEvents: Seq[ObservedCreateEvent],
+      exerciseEvents: Seq[ObservedExerciseEvent],
+  ) {
+    private val _actualTemplateNames =
+      (createEvents.map(_.templateName) ++ exerciseEvents.map(_.templateName)).toSet
+    require(
+      _actualTemplateNames.subsetOf(expectedTemplateNames),
+      s"${_actualTemplateNames} must be a subset of $expectedTemplateNames",
+    )
+
+    private val consumingExercises: Seq[ObservedExerciseEvent] = exerciseEvents.filter(_.consuming)
+
+    val (archiveExercises, nonArchiveConsumingExercises)
+        : (Seq[ObservedExerciseEvent], Seq[ObservedExerciseEvent]) =
+      consumingExercises.partition(_.choiceName == TemplateDescriptor.ArchiveChoiceName)
+
+    val nonconsumingExercises: Seq[ObservedExerciseEvent] = exerciseEvents.filterNot(_.consuming)
+
+    val avgNonArchiveConsumingExerciseSize: Int = {
+      val n = nonArchiveConsumingExercises.size
+      if (n == 0)
+        0
+      else
+        nonArchiveConsumingExercises.map(_.choiceArgumentsSerializedSize).sum / n
+    }
+
+    val avgNonconsumingExerciseSize: Int = {
+      val n = nonconsumingExercises.size
+      if (n == 0)
+        0
+      else
+        nonconsumingExercises.map(_.choiceArgumentsSerializedSize).sum / n
+    }
+
+    val numberOfCreatesPerTemplateName: Map[String, Int] = {
+      val groups = createEvents.groupBy(_.templateName)
+      expectedTemplateNames.map(name => name -> groups.get(name).fold(0)(_.size)).toMap
+    }
+
+    val avgCreateSizePerTemplateName: Map[String, Int] = {
+      val groups = createEvents.groupBy(_.templateName)
+      expectedTemplateNames.map { name =>
+        val avgSize = groups
+          .get(name)
+          .fold(0)(events =>
+            if (events.isEmpty) 0 else events.map(_.createArgumentsSerializedSize).sum / events.size
+          )
+        name -> avgSize
+      }.toMap
+    }
+
+    val numberOfArchivesPerTemplateName: Map[String, Int] = {
+      val groups = archiveExercises
+        .groupBy(_.templateName)
+      expectedTemplateNames.map(name => name -> groups.get(name).fold(0)(_.size)).toMap
+    }
+
+  }
+
+}
+
+/** Collects information about create and exercise events.
+  */
+class EventsObserver(expectedTemplateNames: Set[String], logger: Logger)
+    extends ObserverWithResult[GetTransactionTreesResponse, ObservedEvents](logger) {
+
+  private val createEvents = collection.mutable.ArrayBuffer[ObservedCreateEvent]()
+  private val exerciseEvents = collection.mutable.ArrayBuffer[ObservedExerciseEvent]()
```

Comment:
is Thread (un)safety could be a concern here?

Reply:
Unlikely, these are populated from a sequential stream of transaction tree events

Reply:
Yes, exactly :)